From 9ae823c43fad161431017ba391264b8f14ea493c Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Sun, 02 Feb 2020 10:18:10 +0200
Subject: [PATCH] [SW-8229] goya2: invert revoke bit from flash

When reading from flash or flash_otp revoke bits will
be inversed as default state of flash is set and a write
can only clear the bit.

Change-Id: Id990a79597a071e6c2136bd75b4f00297cc4306a
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index 718142e..0e340a3 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -47,7 +47,6 @@
 
 	switch (current_security_location) {
 	case SPI_FLASH:
-		/* TODO: invert revoke bit logic for spi */
 		spi_flash_get_revoke(&key_revoke);
 		spi_flash_get_hash(key_hash);
 		fw_version_sec = spi_flash_get_fw_version_security();
diff --git a/common/spi_flash.c b/common/spi_flash.c
index 77b601b..73904b7 100644
--- a/common/spi_flash.c
+++ b/common/spi_flash.c
@@ -42,6 +42,11 @@
 	base_writel(val, HL_FLASH_BASE + offset);
 }
 
+static void spi_flash_invert_revoke(uint32_t *key_revoke)
+{
+	*key_revoke = ((~(*key_revoke)) & SECURITY_KEY_REVOKE_MAP);
+}
+
 void spi_flash_get_revoke(uint32_t *key_revoke)
 {
 	uint32_t flash_efuse_offset, revoke_status;
@@ -63,6 +68,8 @@
 	/* check for revoke bit of customer key 4 */
 	if (revoke_status & 1)
 		*key_revoke |= BIT(4);
+
+	spi_flash_invert_revoke(key_revoke);
 }
 
 void spi_flash_get_hash(uint32_t *key_hash)
@@ -126,6 +133,7 @@
 {
 	*key_revoke = spi_flash_readl(FLASH_EFUSE_KEYS_REVOKE_OFFSET) &
 			SECURITY_KEY_REVOKE_MAP;
+	spi_flash_invert_revoke(key_revoke);
 }
 
 void spi_flash_otp_get_hash(uint32_t *key_hash)
